File: 21-Binary-Search-Tree-Part-II/cpp/05-two-sum-iv-input-is-a-bst.cpp
#include <stack>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    void pushAllLeft(TreeNode* node, stack<TreeNode*>& st) {
        while (node != nullptr) {
            st.push(node);
            node = node->left;
        }
    }

    void pushAllRight(TreeNode* node, stack<TreeNode*>& st) {
        while (node != nullptr) {
            st.push(node);
            node = node->right;
        }
    }

public:
    bool findTarget(TreeNode* root, int k) {
        if (!root) return false;

        stack<TreeNode*> leftStack;
        stack<TreeNode*> rightStack;

        pushAllLeft(root, leftStack);
        pushAllRight(root, rightStack);

        TreeNode* leftNode = leftStack.top(); leftStack.pop();
        TreeNode* rightNode = rightStack.top(); rightStack.pop();

        while (leftNode != rightNode) {
            int sum = leftNode->val + rightNode->val;
            if (sum == k) return true;

            if (sum < k) {
                pushAllLeft(leftNode->right, leftStack);
                leftNode = leftStack.top(); leftStack.pop();
            } else {
                pushAllRight(rightNode->left, rightStack);
                rightNode = rightStack.top(); rightStack.pop();
            }
        }

        return false;
    }
};

File: 21-Binary-Search-Tree-Part-II/java/05-TwoSumIvInputIsABst.java
import java.util.*;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean findTarget(TreeNode root, int k) {
        Deque<TreeNode> leftStack = new ArrayDeque<>();
        Deque<TreeNode> rightStack = new ArrayDeque<>();

        pushAllLeft(root, leftStack);
        pushAllRight(root, rightStack);

        TreeNode leftNode = leftStack.pop();
        TreeNode rightNode = rightStack.pop();

        while (leftNode != rightNode) {
            int sum = leftNode.val + rightNode.val;
            if (sum == k) return true;

            if (sum < k) {
                pushAllLeft(leftNode.right, leftStack);
                leftNode = leftStack.pop();
            } else {
                pushAllRight(rightNode.left, rightStack);
                rightNode = rightStack.pop();
            }
        }
        return false;
    }

    private void pushAllLeft(TreeNode node, Deque<TreeNode> stack) {
        while (node != null) {
            stack.push(node);
            node = node.left;
        }
    }

    private void pushAllRight(TreeNode node, Deque<TreeNode> stack) {
        while (node != null) {
            stack.push(node);
            node = node.right;
        }
    }
}
